solutions/serial_server.c

```diff
@@ -26,6 +26,17 @@ int uart_get_char() {
         ch = *REG_PTR(uart_base_vaddr, UARTDR) & RHR_MASK;
     }
 
+    /*
+     * Convert Newline to Carriage return; backspace to DEL
```

Comment:
If the aim is to improve printing text, I think it's better to do these kind of conversion at the output, not the input.

Don't you want to insert a carriage return instead of replacing the newline?
Same for backspace, move back one char and then delete?